tests/base.go

```diff
@@ -0,0 +1,148 @@
+package tests
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/goombaio/namegenerator"
+	"golang.org/x/crypto/ssh"
+)
+
+// UpWg used for up wireguard
+func UpWg(wgConfig string) {
+	f, err := os.Create("/tmp/test.conf")
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer f.Close()
+
+	_, err2 := f.WriteString(wgConfig)
+
+	if err2 != nil {
+		log.Fatal(err2)
+	}
+
+	cmd := exec.Command("sudo", "wg-quick", "up", "/tmp/test.conf")
+	stdout, err := cmd.Output()
+
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	// Print the output
+	fmt.Println(string(stdout))
+}
+
+// DownWG used for down wireguard
+func DownWG() {
+	cmd := exec.Command("sudo", "wg-quick", "down", "/tmp/test.conf")
+	stdout, err := cmd.Output()
+
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	fmt.Println(string(stdout))
+}
+
+// RemoteRun used for ssh host
+func RemoteRun(user string, addr string, cmd string) (string, error) {
+	privateKey := os.Getenv("PRIVATEKEY")
+	key, err := ssh.ParsePrivateKey([]byte(privateKey))
+	if err != nil {
+		return "", err
+	}
+	// Authentication
+	config := &ssh.ClientConfig{
+		User:            user,
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(key),
+		},
+	}
+	// Connect
+	client, err := ssh.Dial("tcp", net.JoinHostPort(addr, "22"), config)
+	if err != nil {
+		return "", err
+	}
+	// Create a session. It is one session per command.
+	session, err := client.NewSession()
+	if err != nil {
+		return "", err
+	}
+	defer session.Close()
+	var b bytes.Buffer  // import "bytes"
+	session.Stdout = &b // get output
+	err = session.Run(cmd)
+	return b.String(), err
+}
+
+func VerifyIPs(wgConfig string, verifyIPs []string) bool {
+	UpWg(wgConfig)
+
+	for i := 0; i < len(verifyIPs); i++ {
+		out, _ := exec.Command("ping", verifyIPs[i], "-c 5", "-i 3", "-w 10").Output()
+		if strings.Contains(string(out), "Destination Host Unreachable") {
+			return false
+		}
+	}
+
+	for i := 0; i < len(verifyIPs); i++ {
+		res, _ := RemoteRun("root", verifyIPs[i], "ifconfig")
+		if !strings.Contains(string(res), verifyIPs[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+func RandomName() string {
+	seed := time.Now().UTC().UnixNano()
+	nameGenerator := namegenerator.NewNameGenerator(seed)
+
+	name := nameGenerator.Generate()
+
+	return name
+}
+
+func Wait(addr string, port string) bool {
```

Comment:
I checked what Dial does under the hood and it operates as follows:
- It sends a TCP packet to the specified address and port
- Then, it resends the packet at exponential intervals (i.e. after 1 second, 1+2 seconds, 1+2+4 seconds and so on).

If the host was down (because it's not up yet or some bug caused it), the sent packets are lost. Once the host becomes up, it responds to the first packet it receives and is considered up.

So making the timeout 60 seconds will cause packets to be sent at these intervals: (1, 3, 7, 15, 31). If the host became up after 35 seconds for example, it won't be detected as up and this will be reported after the 60 seconds are finished.

Retrying 100 times will cause the tests to fail after 100 minutes if the ip is not reachable which looks like a little too much to make sure the host is down. One way to do this is to make the `Wait` method do the retrying and modify the `DialTimeout` to be 12 for example. This way the first packet sent at each `Dial` has some time (12 seconds) to be replied to. And the interval at which packets are sent is not too long because retrying `Dial` in go would reset the interval to 1.

Reply:
The `DialTimeout` duration was changed to 12 seconds. and the retrying logic was changed and it was moved inside the `Wait` function